File: OpenWA/sdk/go/messages.go
package openwa

import "context"

// MessagesService sends and queries messages.
// Backed by src/modules/message/message.controller.ts. NOTE: send paths use the
// "send-" prefix (e.g. /messages/send-text).
type MessagesService struct{ client *Client }

func (s *MessagesService) base(sessionID string) string {
	return "/api/sessions/" + pathEscape(sessionID) + "/messages"
}

// List returns persisted messages for a session.
func (s *MessagesService) List(ctx context.Context, sessionID string, query *ListMessagesQuery) (*MessageListResponse, error) {
	var out MessageListResponse
	err := s.client.do(ctx, "GET", s.base(sessionID), valuesOf(query), nil, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}

// SendText sends a plain text message.
func (s *MessagesService) SendText(ctx context.Context, sessionID string, body SendTextRequest) (*MessageResponse, error) {
	return s.send(ctx, sessionID, "send-text", body)
}

// SendImage sends an image.
func (s *MessagesService) SendImage(ctx context.Context, sessionID string, body SendMediaRequest) (*MessageResponse, error) {
	return s.send(ctx, sessionID, "send-image", body)
}

// SendVideo sends a video.
func (s *MessagesService) SendVideo(ctx context.Context, sessionID string, body SendMediaRequest) (*MessageResponse, error) {
	return s.send(ctx, sessionID, "send-video", body)
}

// SendAudio sends audio (set PTT for a voice note).
func (s *MessagesService) SendAudio(ctx context.Context, sessionID string, body SendAudioRequest) (*MessageResponse, error) {
	return s.send(ctx, sessionID, "send-audio", body)
}

// SendDocument sends a document.
func (s *MessagesService) SendDocument(ctx context.Context, sessionID string, body SendMediaRequest) (*MessageResponse, error) {
	return s.send(ctx, sessionID, "send-document", body)
}

// SendSticker sends a sticker.
func (s *MessagesService) SendSticker(ctx context.Context, sessionID string, body SendMediaRequest) (*MessageResponse, error) {
	return s.send(ctx, sessionID, "send-sticker", body)
}

// SendLocation sends a location pin.
func (s *MessagesService) SendLocation(ctx context.Context, sessionID string, body SendLocationRequest) (*MessageResponse, error) {
	return s.send(ctx, sessionID, "send-location", body)
}

// SendContact sends a contact card.
func (s *MessagesService) SendContact(ctx context.Context, sessionID string, body SendContactRequest) (*MessageResponse, error) {
	return s.send(ctx, sessionID, "send-contact", body)
}

// SendTemplate sends a stored template.
func (s *MessagesService) SendTemplate(ctx context.Context, sessionID string, body SendTemplateRequest) (*MessageResponse, error) {
	return s.send(ctx, sessionID, "send-template", body)
}

// SendPoll sends a native WhatsApp poll (2–12 options).
func (s *MessagesService) SendPoll(ctx context.Context, sessionID string, body SendPollRequest) (*MessageResponse, error) {
	return s.send(ctx, sessionID, "send-poll", body)
}

func (s *MessagesService) send(ctx context.Context, sessionID, segment string, body any) (*MessageResponse, error) {
	var out MessageResponse
	err := s.client.do(ctx, "POST", s.base(sessionID)+"/"+segment, nil, body, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}

// Reply replies to a quoted message.
func (s *MessagesService) Reply(ctx context.Context, sessionID string, body ReplyMessageRequest) (*MessageResponse, error) {
	return s.send(ctx, sessionID, "reply", body)
}

// Forward forwards a message between chats.
func (s *MessagesService) Forward(ctx context.Context, sessionID string, body ForwardMessageRequest) (*MessageResponse, error) {
	return s.send(ctx, sessionID, "forward", body)
}

// React adds (or, with an empty emoji, removes) a reaction.
func (s *MessagesService) React(ctx context.Context, sessionID string, body ReactMessageRequest) (*SuccessResult, error) {
	var out SuccessResult
	err := s.client.do(ctx, "POST", s.base(sessionID)+"/react", nil, body, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}

// Delete deletes a message.
func (s *MessagesService) Delete(ctx context.Context, sessionID string, body DeleteMessageRequest) (*SuccessResult, error) {
	var out SuccessResult
	err := s.client.do(ctx, "POST", s.base(sessionID)+"/delete", nil, body, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}

// EditMessage edits the text of a message sent by this account. Requires an
// OPERATOR-level key. The server responds 404 when the message is not found.
func (s *MessagesService) EditMessage(ctx context.Context, sessionID string, body EditMessageRequest) (*MessageResponse, error) {
	var out MessageResponse
	err := s.client.do(ctx, "POST", s.base(sessionID)+"/edit", nil, body, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}

// History reads live chat history from WhatsApp.
func (s *MessagesService) History(ctx context.Context, sessionID, chatID string, query *MessageHistoryQuery) ([]ChatHistoryMessage, error) {
	var out []ChatHistoryMessage
	path := s.base(sessionID) + "/" + pathEscape(chatID) + "/history"
	err := s.client.do(ctx, "GET", path, valuesOf(query), nil, &out)
	return out, err
}

// Reactions returns the reactions on a message.
func (s *MessagesService) Reactions(ctx context.Context, sessionID, chatID, messageID string) ([]ReactionRecord, error) {
	var out []ReactionRecord
	path := s.base(sessionID) + "/" + pathEscape(chatID) + "/" + pathEscape(messageID) + "/reactions"
	err := s.client.do(ctx, "GET", path, nil, nil, &out)
	return out, err
}

// Pin pins a message in its chat for a bounded window.
func (s *MessagesService) Pin(ctx context.Context, sessionID string, body PinMessageRequest) (*SuccessResult, error) {
	var out SuccessResult
	if err := s.client.do(ctx, "POST", s.base(sessionID)+"/pin", nil, body, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// VotePoll casts a vote on a poll. Not supported on the Baileys engine (501).
func (s *MessagesService) VotePoll(ctx context.Context, sessionID string, body VotePollRequest) (*SuccessResult, error) {
	var out SuccessResult
	if err := s.client.do(ctx, "POST", s.base(sessionID)+"/vote-poll", nil, body, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Star stars or unstars a message.
func (s *MessagesService) Star(ctx context.Context, sessionID string, body StarMessageRequest) (*SuccessResult, error) {
	var out SuccessResult
	if err := s.client.do(ctx, "POST", s.base(sessionID)+"/star", nil, body, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Unpin removes a message's pin.
func (s *MessagesService) Unpin(ctx context.Context, sessionID string, body UnpinMessageRequest) (*SuccessResult, error) {
	var out SuccessResult
	if err := s.client.do(ctx, "POST", s.base(sessionID)+"/unpin", nil, body, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Media fetches a message's archived media bytes. The server answers 404 when
// nothing is archived for the message (archiving off when it arrived, no media,
// over the archive cap, or cleared by retention).
func (s *MessagesService) Media(ctx context.Context, sessionID, chatID, messageID string) (*MessageMedia, error) {
	path := s.base(sessionID) + "/" + pathEscape(chatID) + "/" + pathEscape(messageID) + "/media"
	data, contentType, err := s.client.doRaw(ctx, "GET", path, nil, nil)
	if err != nil {
		return nil, err
	}
	return &MessageMedia{Data: data, ContentType: contentType}, nil
}

// SendBulk queues a batch of messages.
func (s *MessagesService) SendBulk(ctx context.Context, sessionID string, body SendBulkRequest) (*BulkMessageResponse, error) {
	var out BulkMessageResponse
	err := s.client.do(ctx, "POST", s.base(sessionID)+"/send-bulk", nil, body, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}

// BatchStatus returns the status of a bulk batch.
func (s *MessagesService) BatchStatus(ctx context.Context, sessionID, batchID string) (*BatchStatusResponse, error) {
	var out BatchStatusResponse
	path := s.base(sessionID) + "/batch/" + pathEscape(batchID)
	err := s.client.do(ctx, "GET", path, nil, nil, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}

// CancelBatch cancels a running batch. Requires an OPERATOR-level key.
func (s *MessagesService) CancelBatch(ctx context.Context, sessionID, batchID string) (*BatchStatusResponse, error) {
	var out BatchStatusResponse
	path := s.base(sessionID) + "/batch/" + pathEscape(batchID) + "/cancel"
	err := s.client.do(ctx, "POST", path, nil, nil, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}

File: OpenWA/sdk/python/openwa/resources/messages.py
"""Messages resource — sending and querying messages.

Backed by ``src/modules/message/message.controller.ts``.
NOTE: the real paths use the ``/send-`` prefix, e.g. ``/messages/send-text``.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from .._http import quote_segment
from ..types import (
    MessageMedia,
    PinMessageRequest,
    StarMessageRequest,
    VotePollRequest,
    UnpinMessageRequest,
    BatchStatusResponse,
    BulkMessageResponse,
    ChatHistoryMessage,
    DeleteMessageRequest,
    EditMessageRequest,
    ForwardMessageRequest,
    ListMessagesQuery,
    MessageHistoryQuery,
    MessageListResponse,
    MessageResponse,
    ReactionRecord,
    ReactMessageRequest,
    ReplyMessageRequest,
    SendBulkRequest,
    SendContactRequest,
    SendLocationRequest,
    SendAudioRequest,
    SendMediaRequest,
    SendPollRequest,
    SendTemplateRequest,
    SendTextRequest,
    SuccessResult,
)

if TYPE_CHECKING:
    from .._http import HttpExecutor


class MessagesResource:
    def __init__(self, http: "HttpExecutor") -> None:
        self._http = http

    def list(self, session_id: str, query: ListMessagesQuery | None = None) -> MessageListResponse:
        return self._http.request("GET", f"/api/sessions/{quote_segment(session_id)}/messages", query=query)

    def send_text(self, session_id: str, body: SendTextRequest) -> MessageResponse:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/send-text", body=body)

    def send_image(self, session_id: str, body: SendMediaRequest) -> MessageResponse:
        return self._send_media(session_id, "send-image", body)

    def send_video(self, session_id: str, body: SendMediaRequest) -> MessageResponse:
        return self._send_media(session_id, "send-video", body)

    def send_audio(self, session_id: str, body: SendAudioRequest) -> MessageResponse:
        return self._send_media(session_id, "send-audio", body)

    def send_document(self, session_id: str, body: SendMediaRequest) -> MessageResponse:
        return self._send_media(session_id, "send-document", body)

    def send_sticker(self, session_id: str, body: SendMediaRequest) -> MessageResponse:
        return self._send_media(session_id, "send-sticker", body)

    def _send_media(self, session_id: str, segment: str, body: SendMediaRequest) -> MessageResponse:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/{quote_segment(segment)}", body=body)

    def send_location(self, session_id: str, body: SendLocationRequest) -> MessageResponse:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/send-location", body=body)

    def send_contact(self, session_id: str, body: SendContactRequest) -> MessageResponse:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/send-contact", body=body)

    def send_template(self, session_id: str, body: SendTemplateRequest) -> MessageResponse:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/send-template", body=body)

    def send_poll(self, session_id: str, body: SendPollRequest) -> MessageResponse:
        """Send a native WhatsApp poll (2–12 options)."""
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/send-poll", body=body)

    def reply(self, session_id: str, body: ReplyMessageRequest) -> MessageResponse:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/reply", body=body)

    def forward(self, session_id: str, body: ForwardMessageRequest) -> MessageResponse:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/forward", body=body)

    def react(self, session_id: str, body: ReactMessageRequest) -> SuccessResult:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/react", body=body)

    def delete(self, session_id: str, body: DeleteMessageRequest) -> SuccessResult:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/delete", body=body)

    def edit_message(self, session_id: str, body: EditMessageRequest) -> MessageResponse:
        """Edit the text of a message sent by this account. 404 when the message is not found."""
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/edit", body=body)

    def history(
        self, session_id: str, chat_id: str, query: MessageHistoryQuery | None = None
    ) -> list[ChatHistoryMessage]:
        return self._http.request(
            "GET", f"/api/sessions/{quote_segment(session_id)}/messages/{quote_segment(chat_id)}/history", query=query
        )

    def reactions(self, session_id: str, chat_id: str, message_id: str) -> list[ReactionRecord]:
        return self._http.request(
            "GET", f"/api/sessions/{quote_segment(session_id)}/messages/{quote_segment(chat_id)}/{quote_segment(message_id)}/reactions"
        )

    def pin(self, session_id: str, body: PinMessageRequest) -> SuccessResult:
        """Pin a message. ``durationSeconds`` must be 86400, 604800 or 2592000; defaults to 24h."""
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/pin", body=body)

    def vote_poll(self, session_id: str, body: VotePollRequest) -> SuccessResult:
        """Cast a vote on a poll. Not supported on the Baileys engine (501)."""
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/vote-poll", body=body)

    def star(self, session_id: str, body: StarMessageRequest) -> SuccessResult:
        """Star or unstar a message. Best-effort on whatsapp-web.js."""
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/star", body=body)

    def unpin(self, session_id: str, body: UnpinMessageRequest) -> SuccessResult:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/unpin", body=body)

    def media(self, session_id: str, chat_id: str, message_id: str) -> MessageMedia:
        """Fetch a message's archived media bytes (404 when nothing is archived for it)."""
        data, content_type = self._http.request_bytes(
            "GET",
            f"/api/sessions/{quote_segment(session_id)}/messages/{quote_segment(chat_id)}/{quote_segment(message_id)}/media",
        )
        return {"data": data, "contentType": content_type}

    def send_bulk(self, session_id: str, body: SendBulkRequest) -> BulkMessageResponse:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/send-bulk", body=body)

    def batch_status(self, session_id: str, batch_id: str) -> BatchStatusResponse:
        return self._http.request("GET", f"/api/sessions/{quote_segment(session_id)}/messages/batch/{quote_segment(batch_id)}")

    def cancel_batch(self, session_id: str, batch_id: str) -> BatchStatusResponse:
        """Cancel a running batch. Requires an OPERATOR-level key."""
        return self._http.request(
            "POST", f"/api/sessions/{quote_segment(session_id)}/messages/batch/{quote_segment(batch_id)}/cancel"
        )
